src/System.Text.Json/tests/JsonDocumentTests.cs

```diff
@@ -3712,6 +3712,43 @@ private static void AssertContents(string expectedValue, ArrayBufferWriter<byte>
 #endif
                     ));
         }
+
+        [Fact]
+        // Issue #39930
+        public static void Disposing_On_Multiple_Threads_Should_Not_Return_To_ArrayPool_Twice()
+        {
+            Action<object> disposeAction = (object document) => ((JsonDocument)document).Dispose();
+
+            // Create a bunch of parallel tasks that call Dispose several times on the same object.
+            Task[] tasks = new Task[100];
+            int count = 0;
+            for (int j = 0; j < 10; j++)
+            {
+                JsonDocument document = JsonDocument.Parse("123" + j);
+                for (int i = 0; i < 10; i++)
+                {
+                    tasks[count] = new Task(disposeAction, document);
+                    tasks[count].Start();
+
+                    count++;
+                }
+            }
+
+            Task.WaitAll(tasks);
+
+            // When ArrayPool gets corrupted, the Rent method might pull an already rented array, which is incorrect.
+            // So we will pull as many arrays as calls to JsonElement.Dispose and check they are unique.
+            HashSet<IntPtr> uniqueAddresses = new HashSet<IntPtr>();
+            while (count > 0)
+            {
+                byte[] arr = ArrayPool<byte>.Shared.Rent(4);
+                GCHandle pinned = GCHandle.Alloc(arr, GCHandleType.Pinned);
```

Comment:
You really should unpin all of these :).

But this is overly complicated; it just needs to do a HashSet of `byte[]`, since reference equality is all we're after.

Reply:
> But this is overly complicated; it just needs to do a HashSet of `byte[]`, since reference equality is all we're after.

Good point, agreed.

Reply:
> You really should unpin all of these :).

One might ask, Why?

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle?view=netframework-4.8#remarks

> When the handle goes out of scope you must explicitly release it by calling the Free method; otherwise, memory leaks may occur. When you free a pinned handle, the associated object will be unpinned and will become eligible for garbage collection, if there are no other references to it.

Basically, the GC will not collect the resource otherwise.

Reply:
Forgot to copy/paste that line. 
Ok, I am taking your suggestion @bartonjs, thanks.